### PR TITLE
Added ipydatetime

### PR DESCRIPTION
**Pull request by thewchan**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by thewchan** [comment]
@conda-forge/staged-recipes Hey folks, this one is kinda a weird one (for me anyway). It is a jupyter extension, but it's npm package is a different name than the one in PyPI, so the test gets confused I think with the import test (it can't find the package under the name in PyPI I think?). I've tried tinkering with the path in `build.sh` to no avail. For now I've removed the import test in `meta.yml` and everything else I think works. Please let me know if I did this correctly. Thanks!

**Comment by ocefpaf** [comment]
@bollwyvl you have some experiences with JL packages, right? Do you mind taking a look at this one?

File: recipes/ipydatetime/build.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+set -ex
+
+$PYTHON -m pip install . -vv
+npm pack jupyter-widget-datetime@${PKG_VERSION}
```
Comment (bollwyvl) [inline recipes/ipydatetime/build.sh:5]: the `sdist` from pypi includes this tarball already, and will already be deployed to the correct location from with `pip install`. I recommend dropping this altogether, as then this can be a proper `noarch: python` package.

File: recipes/ipydatetime/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
+{% set name = "ipydatetime" %}
+{% set version = "1.2.0" %}
+
+package:
+  name: {{ name|lower }}
```
Comment (bollwyvl) [inline recipes/ipydatetime/meta.yaml:5]: this can be `noarch: python`

File: recipes/ipydatetime/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  sha256: 9b083e08efef140c29091fe114ea2921f88989e56f0b5eecdde9ac0157bb0b10
+
+build:
+  number: 0
+
+requirements:
+  host:
+    - jupyter-packaging
+    - nodejs
```
Comment (bollwyvl) [inline recipes/ipydatetime/meta.yaml:18]: does not need `nodejs` in (at least) `host`

File: recipes/ipydatetime/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    - jupyterlab >=2
+    - nodejs
+    - python
+    - tzlocal
+
+test:
+  commands:
+    - pip check
+  requires:
+    - pip
```
Comment (bollwyvl) [inline recipes/ipydatetime/meta.yaml:33]: it's worth adding a check for whether the labextension is detected, see https://conda-forge.org/docs/maintainer/knowledge_base.html#jupyterlab-extensions

File: recipes/ipydatetime/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+requirements:
+  host:
+    - jupyter-packaging
+    - nodejs
+    - python
+    - pip
+  run:
+    - ipywidgets >=7.0.0
+    - jupyter-packaging
```
Comment (bollwyvl) [inline recipes/ipydatetime/meta.yaml:23]: it probably doesn't need `jupyter-packaging` to run

**Comment by thewchan** [comment]
@bollwyvl 
Thanks so much for the suggestions; I have made the edits/commits and we'll see how the tests go.

**Comment by thewchan** [comment]
@bollwyvl @ocefpaf 

After making the recommended changes we're running into `ERROR: Could not build wheels for ipydatetime which use PEP 517 and cannot be installed directly`

I saw this when I first tried to build this as no arch, and changing to use the build.sh script and adding nodejs (one or the other, or both) "fixed" this but raised other issues. Thoughts?

File: recipes/asgi-csrf/LICENSE [hunk 6]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (bollwyvl) [inline recipes/asgi-csrf/LICENSE:1]: can you get this other recipe out of here? seems irrelevant?

File: recipes/ipydatetime/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  version: {{ version }}
+
+source:
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  sha256: 9b083e08efef140c29091fe114ea2921f88989e56f0b5eecdde9ac0157bb0b10
+
+build:
+  noarch: python
+  number: 0
+  script: {{ PYTHON }} -m pip install . -vv
```
Comment (bollwyvl) [inline recipes/ipydatetime/meta.yaml:15]: try deleting `pyproject.toml`:

```yaml
script: 
- {{ PYTHON }} -c '__import__("os").unlink("pyproject.toml")'
- {{ PYTHON }} -m pip install . -vv --install-option="--skip-npm"
```

File: recipes/asgi-csrf/LICENSE (hunk 6 above)
Comment (thewchan) [reply recipes/asgi-csrf/LICENSE:1]: yeah looks like I got confused managing my local branch

**Comment by thewchan** [comment]
Similar errors:

```
ValueError: missing files: ['/tmp/pip-req-build-0x5ge7yt/ts/lib/plugin.js']
  Building wheel for ipydatetime (PEP 517): finished with status 'error'
  ERROR: Failed building wheel for ipydatetime
Failed to build ipydatetime
ERROR: Could not build wheels for ipydatetime which use PEP 517 and cannot be installed directly
```

**Comment by thewchan** [comment]
It might need nodejs after all?

File: recipes/ipydatetime/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+source:
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  sha256: 9b083e08efef140c29091fe114ea2921f88989e56f0b5eecdde9ac0157bb0b10
+
+build:
+  noarch: python
+  number: 0
+  script:
+    - {{ PYTHON }} -m pip install . -vv
+    - {{ PYTHON }} -c '__import__("os").unlink("pyproject.toml")'
```
Comment (bollwyvl) [inline recipes/ipydatetime/meta.yaml:17]: invert the order here

Comment (thewchan) [reply recipes/ipydatetime/meta.yaml:17]: gotcha; sorry I missed that

**Comment by thewchan** [comment]
Looks like it failed that jupyter check test:

```
+ pip check
No broken requirements found.
+ jupyter labextension check @jupyterlab/jupyter-widget-datetime
```

File: recipes/ipydatetime/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    - ipywidgets >=7.0.0
+    - jupyterlab >=2
+    - nodejs
+    - python >=3.6
+    - tzlocal
+
+test:
+  commands:
+    - pip check
+    - jupyter labextension check @jupyterlab/jupyter-widget-datetime; echo $?
```
Comment (bollwyvl) [inline recipes/ipydatetime/meta.yaml:34]: try a `jupyter labextension list` as well. Though i don't see the tarball getting copied in the log... will make a suggestion above.

File: recipes/ipydatetime/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  sha256: 9b083e08efef140c29091fe114ea2921f88989e56f0b5eecdde9ac0157bb0b10
+
+build:
+  noarch: python
+  number: 0
+  script:
+    - {{ PYTHON }} -c '__import__("os").unlink("pyproject.toml")'
+    - {{ PYTHON }} -m pip install . -vv
+
```
Comment (bollwyvl) [inline recipes/ipydatetime/meta.yaml:18]: here, i guess add:
```yaml
- {{ PYTHON }} -c "import pathlib; src = pathlib.Path('ts/lab-dist/jupyter-widget-datetime-{{ version }}.tgz'), pathlib.Path('''{{ PREFIX }}'''); dest = pathlib.Path('''{{ PREFIX }}''', '/share/jupyter/labextensions/js', src.name); dest.parent.mkdir(parents=True); src.rename(dest)"
```

might take some tweaking, though...

File: recipes/ipydatetime/meta.yaml (hunk 9 above)
Comment (bollwyvl) [reply recipes/ipydatetime/meta.yaml:34]: also the fully-qualified name is `jupyter-widget-datetime`, and might not show up as "working" unless built, so perhaps:
```
- jupyter lab build
- jupyter labextension list
- jupyter labextension check jupyter-widget-datetime; echo $?
```

**Comment by thewchan** [comment]
More mysteriousness...

```
File "<string>", line 1, in <module>
AttributeError: 'tuple' object has no attribute 'name' subprocess.
CalledProcessError: Command '['/bin/bash', '-o', 'errexit', '/home/conda/staged-recipes/build_artifacts/ipydatetime_1616513521080/work/conda_build.sh']' returned non-zero exit status 1
```

**Comment by bollwyvl** [comment]
Right, that's the tweaking part. If you're more comfortable with a python
script, you can move it there and call that. I can recommend getting a
local environment set up to reduce cycle time...

**Comment by thewchan** [comment]
@conda-forge/staged-recipes This is all building correctly except for the windows one now, anyone have any ideas?

**Comment by thewchan** [comment]
Okay this is FINALLY passing all tests, but I suspect the meta.yaml contains a lot of mangled stuff and would really benefit few extra pairs of eyes going over it. Thanks @conda-forge/staged-recipes !

File: recipes/ipydatetime/meta.yaml (hunk 8 above)
Comment (ocefpaf) [reply recipes/ipydatetime/meta.yaml:16]: Just curios, why keeping the pyproject.toml did not work here?

Comment (thewchan) [reply recipes/ipydatetime/meta.yaml:16]: @bollwyvl will have to answer that; I was just blindly doing everything he was telling me to do

File: recipes/ipydatetime/meta.yaml (hunk 3 above)
Comment (isuruf) [reply recipes/ipydatetime/meta.yaml:28]: `nodejs` should only be in `build`.

Comment (thewchan) [reply recipes/ipydatetime/meta.yaml:28]: Do you mean I should remove it from both `host` and `run`?

Comment (isuruf) [reply recipes/ipydatetime/meta.yaml:28]: yes

Comment (thewchan) [reply recipes/ipydatetime/meta.yaml:28]: Ok, let me try that

File: recipes/ipydatetime/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    - {{ PYTHON }} -c '__import__("os").unlink("pyproject.toml")'
+    - {{ PYTHON }} -m pip install . -vv
+    - npm pack jupyter-widget-datetime@${PKG_VERSION}  # [unix]
+    - npm pack jupyter-widget-datetime@%PKG_VERSION%  # [win]
+    - mkdir -p ${PREFIX}/share/jupyter/lab/extensions/js  # [unix]
+    - mkdir -p %PREFIX%/share/jupyter/lab/extensions/js  # [win]
+    - cp jupyter-widget-datetime-${PKG_VERSION}.tgz ${PREFIX}/share/jupyter/lab/extensions/js  # [unix]
+    - cp jupyter-widget-datetime-%PKG_VERSION%.tgz ${PREFIX}/share/jupyter/lab/extensions/js  # [win]
+
+requirements:
```
Comment (isuruf) [inline recipes/ipydatetime/meta.yaml:25]:
```suggestion
requirements:
  build:
    - nodejs
```

Comment (thewchan) [reply recipes/ipydatetime/meta.yaml:25]: @isuruf Done!
